src/eval/str.rs

```diff
@@ -257,16 +259,75 @@ impl Str {
     }
 
     /// Replace at most `count` occurrences of the given pattern with a
-    /// replacement string (beginning from the start).
-    pub fn replace(&self, pattern: StrPattern, with: Self, count: Option<usize>) -> Self {
-        match pattern {
-            StrPattern::Str(pat) => match count {
-                Some(n) => self.0.replacen(pat.as_str(), &with, n).into(),
-                None => self.0.replace(pat.as_str(), &with).into(),
+    /// replacement string or function (beginning from the start). If no count is given,
+    /// all occurences are replaced
+    pub fn replace(
+        &self,
+        vm: &mut Vm,
+        span: Span,
+        pattern: StrPattern,
+        with: Replacement,
+        count: Option<usize>,
+    ) -> SourceResult<Self> {
+        match with {
+            Replacement::Func(func) => match &pattern {
+                StrPattern::Str(pat) => {
+                    // call `func` with the pattern string and use the output to either replace
+                    // n many occurences or all
+                    let args =
+                        Args::new(func.span(), [Value::Str(Str::from(pat.as_str()))]);
+                    let res = func.call_vm(vm, args)?.cast::<Str>().at(func.span())?;
+                    let new = if let Some(n) = count {
+                        self.0.replacen(pat.as_str(), res.as_str(), n)
+                    } else {
+                        self.0.replace(pat.as_str(), res.as_str())
+                    };
+                    Ok(new.into())
+                }
+                StrPattern::Regex(re) => {
+                    // heuristic: assume the new string is about the same length as the current string
+                    let mut new = String::with_capacity(self.as_str().len());
+                    let mut last_match = 0;
+                    let matches_dicts = re
+                        .captures_iter(self)
+                        .map(captures_to_dict)
+                        .take(count.unwrap_or(usize::MAX));
+                    for match_dict in matches_dicts {
+                        // safety: we can unwrap here since we know due to the definition of `captures_to_dict`
+                        // what the keys are
+                        let start = match_dict
```

Comment:
why convert the captures to dicts and then operate on them here? I haven't looked closely, but probably you should operate on the rust representation in the rust code and then only call captures_to_dict once you're done